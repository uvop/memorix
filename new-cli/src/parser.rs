use crate::{
    create_enum_with_const_slice, impl_from_and_to_sdl_for_enum, impl_from_and_to_sdl_for_struct,
    parser_tools::{indent, FromSdl, ToSdl},
    permutation_many,
};
use nom::{
    branch::alt,
    bytes::complete::{is_not, tag},
    character::complete::{alphanumeric1, char, multispace0, multispace1},
    combinator::{cut, map, opt, value, verify},
    error::{context, ParseError},
    multi::{many0, many_m_n, separated_list1},
    sequence::{delimited, pair, preceded, terminated, tuple},
    IResult,
};

use serde::{Deserialize, Serialize};

#[derive(Debug, PartialEq, Clone, Serialize, Deserialize)]
pub struct BracketString(String);

impl BracketString {
    pub fn to_string(&self) -> String {
        self.0.clone()
    }
}

impl FromSdl for BracketString {
    fn from_sdl<'a, E: ParseError<&'a str> + nom::error::ContextError<&'a str>>(
        input: &'a str,
    ) -> IResult<&'a str, Self, E>
    where
        Self: Sized,
    {
        map(delimited(char('"'), is_not("\""), char('"')), |x: &str| {
            BracketString(x.to_string())
        })(input)
    }
}

impl ToSdl for BracketString {
    fn to_sdl(&self, _: usize) -> String {
        format!("\"{}\"", self.0).to_string()
    }
}

#[derive(Debug, PartialEq, Clone, Serialize, Deserialize)]
pub struct Schema {
    pub config: Option<Config>,
    pub global_namespace: Namespace,
    pub namespaces: Vec<(String, Namespace)>,
}

#[derive(Debug, PartialEq, Clone, Serialize, Deserialize)]
pub struct Config {
    pub import: Option<Vec<BracketString>>,
    pub export: Option<Export>,
}

impl_from_and_to_sdl_for_struct! {
    (Config, 2),
    (import: Vec<BracketString>, 0, false),
    (export: Export, 1, false),
}

#[derive(Debug, PartialEq, Clone, Serialize, Deserialize)]
pub struct Export {
    pub engine: Engine,
    pub files: Option<Vec<FileConfig>>,
}
impl_from_and_to_sdl_for_struct! {
    (Export, 2),
    (engine: Engine, 0, true),
    (files: Vec<FileConfig>, 1, false),
}

#[derive(Debug, PartialEq, Clone, Serialize, Deserialize)]
pub enum Engine {
    Redis(Value),
}

#[derive(Debug, PartialEq, Clone, Serialize, Deserialize)]
pub struct FileConfig {
    pub language: Language,
    pub path: BracketString,
}
impl_from_and_to_sdl_for_struct! {
    (FileConfig, 2),
    (language: Language, 0, true),
    (path: BracketString, 1, true),
}

#[derive(Debug, PartialEq, Clone, Serialize, Deserialize)]
pub enum Language {
    TypeScript,
    Python,
    Rust,
}

impl_from_and_to_sdl_for_enum!(
    Language,
    TypeScript => "typescript",
    Python => "python",
    Rust => "rust"
);

#[derive(Debug, PartialEq, Clone, Serialize, Deserialize)]
pub struct Namespace {
    pub defaults: Option<NamespaceDefaults>,
    pub enum_items: Option<EnumItems>,
    pub type_items: Option<Vec<(String, TypeItem)>>,
    pub cache_items: Option<Vec<(String, CacheItem)>>,
    pub pubsub_items: Option<Vec<(String, PubSubItem)>>,
    pub task_items: Option<Vec<(String, TaskItem)>>,
}

#[derive(Debug, PartialEq, Clone, Serialize, Deserialize)]
pub struct ItemWithPublic<O> {
    pub public: Vec<O>,
}

#[derive(Debug, PartialEq, Clone, Serialize, Deserialize)]
pub struct NamespaceDefaults {
    pub cache_ttl: Option<Value>,
    pub task_queue_type: Option<Value>,
}
impl_from_and_to_sdl_for_struct! {
    (NamespaceDefaults, 2),
    (cache_ttl: Value, 0, false),
    (task_queue_type: Value, 1, false),
}

#[derive(Debug, PartialEq, Clone, Serialize, Deserialize)]
pub struct EnumItems {
    pub items: Vec<EnumItem>,
}

#[derive(Debug, PartialEq, Clone, Serialize, Deserialize)]
pub struct EnumItem {
    pub name: String,
    pub values: Vec<String>,
}

#[derive(Debug, PartialEq, Clone, Serialize, Deserialize)]
pub struct CacheItem {
    pub key: Option<TypeItem>,
    pub payload: TypeItem,
    pub ttl: Option<Value>,
    pub public: Option<Vec<CacheOperation>>,
}
impl_from_and_to_sdl_for_struct! {
    (CacheItem, 4),
    (key: TypeItem, 0, false),
    (payload: TypeItem, 1, true),
    (ttl: Value, 2, false),
    (public: Vec<CacheOperation>, 3, false),
}

#[derive(Debug, PartialEq, Clone, Serialize, Deserialize)]
pub struct PubSubItem {
    pub key: Option<TypeItem>,
    pub payload: TypeItem,
    pub public: Option<Vec<PubSubOperation>>,
}
impl_from_and_to_sdl_for_struct! {
    (PubSubItem, 3),
    (key: TypeItem, 0, false),
    (payload: TypeItem, 1, true),
    (public: Vec<PubSubOperation>, 2, false),
}

#[derive(Debug, PartialEq, Clone, Serialize, Deserialize)]
pub struct TaskItem {
    pub key: Option<TypeItem>,
    pub payload: TypeItem,
    pub queue_type: Option<Value>,
    pub public: Option<Vec<TaskOperation>>,
}
impl_from_and_to_sdl_for_struct! {
    (TaskItem, 4),
    (key: TypeItem, 0, false),
    (payload: TypeItem, 1, true),
    (queue_type: Value, 2, false),
    (public: Vec<TaskOperation>, 3, false),
}

#[derive(Debug, PartialEq, Clone, Serialize, Deserialize)]
pub enum Value {
    String(String),
    Env(String),
}

#[derive(Debug, PartialEq, Clone, Serialize, Deserialize)]
pub enum TypeItem {
    U32,
    I32,
    U64,
    I64,
    F32,
    F64,
    String,
    Boolean,
    Optional(Box<TypeItem>),
    Array(Box<TypeItem>),
    Object(Vec<(String, TypeItem)>),
    Reference(String),
}

create_enum_with_const_slice! {
#[derive(Debug, PartialEq, Clone, Serialize, Deserialize)]
pub enum CacheOperation {
    Get,
    Set,
    Delete,
}
pub ALL_CACHE_OPERATIONS
}

impl_from_and_to_sdl_for_enum!(
    CacheOperation,
    Get => "get",
    Set => "set",
    Delete => "delete"
);

create_enum_with_const_slice! {
#[derive(Debug, PartialEq, Clone, Serialize, Deserialize)]
pub enum PubSubOperation {
    Publish,
    Subscribe,
}
pub ALL_PUBSUB_OPERATIONS
}

impl_from_and_to_sdl_for_enum!(
    PubSubOperation,
    Publish => "publish",
    Subscribe => "subscribe",
);

create_enum_with_const_slice! {
#[derive(Debug, PartialEq, Clone, Serialize, Deserialize)]
pub enum TaskOperation {
    Enqueue,
    Dequeue,
    Empty,
    GetLen,
}
pub ALL_TASK_OPERATIONS
}

impl_from_and_to_sdl_for_enum!(
    TaskOperation,
    Enqueue => "enqueue",
    Dequeue => "dequeue",
    Empty => "empty",
    GetLen => "get_len"
);

impl FromSdl for TypeItem {
    fn from_sdl<'a, E: ParseError<&'a str> + nom::error::ContextError<&'a str>>(
        input: &'a str,
    ) -> IResult<&'a str, Self, E>
    where
        Self: Sized,
    {
        map(
            tuple((
                preceded(
                    multispace0,
                    alt((
                        context(
                            "single type config array",
                            map(
                                preceded(
                                    tuple((char('['), multispace0)),
                                    cut(terminated(
                                        TypeItem::from_sdl,
                                        preceded(multispace0, char(']')),
                                    )),
                                ),
                                |v| TypeItem::Array(Box::new(v)),
                            ),
                        ),
                        map(Vec::<(String, TypeItem)>::from_sdl, TypeItem::Object),
                        alt((
                            value(TypeItem::U32, tag("u32")),
                            value(TypeItem::I32, tag("i32")),
                            value(TypeItem::U64, tag("u64")),
                            value(TypeItem::I64, tag("i64")),
                            value(TypeItem::F32, tag("f32")),
                            value(TypeItem::F64, tag("f64")),
                            value(TypeItem::String, tag("string")),
                            value(TypeItem::Boolean, tag("boolean")),
                        )),
                        map(String::from_sdl, TypeItem::Reference),
                    )),
                ),
                opt(tuple((multispace0, char('?')))),
            )),
            |(t, o)| match o.is_some() {
                false => t,
                true => TypeItem::Optional(Box::new(t)),
            },
        )(input)
    }
}

impl ToSdl for TypeItem {
    fn to_sdl(&self, level: usize) -> String {
        match self {
            Self::U32 => "u32".to_string(),
            Self::U64 => "u64".to_string(),
            Self::I32 => "i32".to_string(),
            Self::I64 => "i64".to_string(),
            Self::F32 => "f32".to_string(),
            Self::F64 => "f64".to_string(),
            Self::Boolean => "boolean".to_string(),
            Self::String => "string".to_string(),
            Self::Reference(x) => x.clone(),
            Self::Optional(x) => format!("{}?", x.to_sdl(level),).to_string(),
            Self::Array(x) => format!(
                "[\n{}{}\n{}]",
                indent(level + 1),
                x.to_sdl(level + 1),
                indent(level)
            )
            .to_string(),
            Self::Object(x) => x.to_sdl(level),
        }
    }
}

impl FromSdl for Value {
    fn from_sdl<'a, E: ParseError<&'a str> + nom::error::ContextError<&'a str>>(
        input: &'a str,
    ) -> IResult<&'a str, Self, E>
    where
        Self: Sized,
    {
        alt((
            map(
                preceded(
                    tag("env"),
                    cut(delimited(
                        tuple((multispace0, char('('), multispace0)),
                        String::from_sdl,
                        tuple((multispace0, char(')'))),
                    )),
                ),
                |s| Value::Env(s),
            ),
            map(delimited(char('"'), is_not("\""), char('"')), |x: &str| {
                Value::String(x.to_string())
            }),
        ))(input)
    }
}

impl ToSdl for Value {
    fn to_sdl(&self, _: usize) -> String {
        match self {
            Self::String(x) => format!("\"{}\"", x).to_string(),
            Self::Env(x) => format!("env({})", x).to_string(),
        }
    }
}

impl FromSdl for Engine {
    fn from_sdl<'a, E: ParseError<&'a str> + nom::error::ContextError<&'a str>>(
        input: &'a str,
    ) -> IResult<&'a str, Self, E>
    where
        Self: Sized,
    {
        map(
            preceded(
                tag("Redis("),
                cut(terminated(
                    delimited(multispace0, Value::from_sdl, multispace0),
                    char(')'),
                )),
            ),
            |v| Engine::Redis(v),
        )(input)
    }
}

impl ToSdl for Engine {
    fn to_sdl(&self, level: usize) -> String {
        match self {
            Self::Redis(x) => format!("Redis({})", x.to_sdl(level)).to_string(),
        }
    }
}

impl FromSdl for EnumItems {
    fn from_sdl<'a, E: ParseError<&'a str> + nom::error::ContextError<&'a str>>(
        input: &'a str,
    ) -> IResult<&'a str, Self, E>
    where
        Self: Sized,
    {
        map(
            delimited(
                tuple((multispace0, char('{'))),
                many0(EnumItem::from_sdl),
                tuple((multispace0, char('}'))),
            ),
            |items| EnumItems { items },
        )(input)
    }
}

impl ToSdl for EnumItems {
    fn to_sdl(&self, level: usize) -> String {
        let level_indent = indent(level);
        let mut result = "{\n".to_string();

        for item in &self.items {
            result.push_str(&format!("{}", item.to_sdl(level + 1)));
        }
        result.push_str(&format!("{}}}", level_indent));

        result
    }
}

impl FromSdl for EnumItem {
    fn from_sdl<'a, E: ParseError<&'a str> + nom::error::ContextError<&'a str>>(
        input: &'a str,
    ) -> IResult<&'a str, Self, E>
    where
        Self: Sized,
    {
        map(
            pair(
                preceded(multispace0, String::from_sdl),
                cut(delimited(
                    tuple((multispace0, char('{'), multispace0)),
                    separated_list1(multispace1, alphanumeric1),
                    tuple((multispace0, char('}'))),
                )),
            ),
            |(name, values)| Self {
                name,
                values: values.into_iter().map(|x| x.to_string()).collect(),
            },
        )(input)
    }
}

impl ToSdl for EnumItem {
    fn to_sdl(&self, level: usize) -> String {
        let level_indent = indent(level);
        let next_level_indent = indent(level + 1);
        let mut result = format!("{}{} {{\n", level_indent, self.name).to_string();

        for value in &self.values {
            result.push_str(&format!("{}{}\n", next_level_indent, value));
        }
        result.push_str(&format!("{}}}\n", level_indent));

        result
    }
}

impl FromSdl for Namespace {
    fn from_sdl<'a, E: ParseError<&'a str> + nom::error::ContextError<&'a str>>(
        input: &'a str,
    ) -> IResult<&'a str, Self, E>
    where
        Self: Sized,
    {
        map(
            permutation_many!(
                6,
                (
                    preceded(
                        tuple((multispace0, tag("NamespaceDefaults"), multispace0)),
                        cut(NamespaceDefaults::from_sdl),
                    ),
                    0,
                    false
                ),
                (
                    preceded(
                        tuple((multispace0, tag("Type"), multispace0)),
                        cut(Vec::<(String, TypeItem)>::from_sdl),
                    ),
                    1,
                    false
                ),
                (
                    preceded(
                        tuple((multispace0, tag("Enum"), multispace0)),
                        cut(EnumItems::from_sdl),
                    ),
                    2,
                    false
                ),
                (
                    preceded(
                        tuple((multispace0, tag("Cache"), multispace0)),
                        cut(Vec::<(String, CacheItem)>::from_sdl),
                    ),
                    3,
                    false
                ),
                (
                    preceded(
                        tuple((multispace0, tag("PubSub"), multispace0)),
                        cut(Vec::<(String, PubSubItem)>::from_sdl),
                    ),
                    4,
                    false
                ),
                (
                    preceded(
                        tuple((multispace0, tag("Task"), multispace0)),
                        cut(Vec::<(String, TaskItem)>::from_sdl),
                    ),
                    5,
                    false
                ),
            ),
            |(defaults, type_items, enum_items, cache_items, pubsub_items, task_items)| Namespace {
                defaults,
                type_items,
                enum_items,
                cache_items,
                pubsub_items,
                task_items,
            },
        )(input)
    }
}

impl ToSdl for Namespace {
    fn to_sdl(&self, level: usize) -> String {
        let level_indent = indent(level);
        let mut result = String::from("");

        if let Some(x) = &self.defaults {
            result.push_str(&format!(
                "{}NamespaceDefaults {}\n",
                level_indent,
                x.to_sdl(level + 1)
            ));
        }
        if let Some(x) = &self.type_items {
            result.push_str(&format!("{}Type {}\n", level_indent, x.to_sdl(level)));
        }
        if let Some(x) = &self.enum_items {
            result.push_str(&format!("{}Enum {}\n", level_indent, x.to_sdl(level)));
        }
        if let Some(x) = &self.cache_items {
            result.push_str(&format!("{}Cache {}\n", level_indent, x.to_sdl(level)));
        }
        if let Some(x) = &self.pubsub_items {
            result.push_str(&format!("{}PubSub {}\n", level_indent, x.to_sdl(level)));
        }
        if let Some(x) = &self.task_items {
            result.push_str(&format!("{}Task {}\n", level_indent, x.to_sdl(level)));
        }

        result
    }
}

fn namespaces_from_sdl<'a, E: ParseError<&'a str> + nom::error::ContextError<&'a str>>(
    input: &'a str,
) -> IResult<&'a str, Vec<(String, Namespace)>, E> {
    many0(preceded(
        tuple((multispace0, tag("Namespace"), multispace1)),
        cut(terminated(
            pair(
                terminated(String::from_sdl, tuple((multispace0, char('{')))),
                Namespace::from_sdl,
            ),
            tuple((multispace0, char('}'))),
        )),
    ))(input)
}

fn namespaces_to_sdl(namespaces: &[(String, Namespace)], level: usize) -> String {
    let level_indent = indent(level);
    let mut result = String::from("");
    for (name, namespace) in namespaces {
        result.push_str(&format!(
            "{}Namespace {} {{\n{}{}}}\n\n",
            level_indent,
            name,
            namespace.to_sdl(level + 1),
            level_indent,
        ));
    }
    result
}

impl FromSdl for Schema {
    fn from_sdl<'a, E: ParseError<&'a str> + nom::error::ContextError<&'a str>>(
        input: &'a str,
    ) -> IResult<&'a str, Self, E>
    where
        Self: Sized,
    {
        map(
            tuple((
                opt(preceded(
                    tuple((multispace0, tag("Config"), multispace0)),
                    cut(Config::from_sdl),
                )),
                namespaces_from_sdl,
                opt(Namespace::from_sdl),
                namespaces_from_sdl,
            )),
            |(config, named_namespaces, global_namespace, more_named_namespaces)| Schema {
                config,
                global_namespace: global_namespace.unwrap_or(Namespace {
                    defaults: None,
                    type_items: None,
                    enum_items: None,
                    cache_items: None,
                    pubsub_items: None,
                    task_items: None,
                }),
                namespaces: named_namespaces
                    .into_iter()
                    .chain(more_named_namespaces.into_iter())
                    .collect(),
            },
        )(input)
    }
}

pub fn schema_to_sdl(schema: &Schema) -> String {
    let mut result = String::from("");
    let level = 0;

    if let Some(x) = &schema.config {
        result.push_str(&format!("Config {}\n\n", x.to_sdl(level)));
    }
    result.push_str(&format!("{}\n", schema.global_namespace.to_sdl(level)));
    result.push_str(&format!(
        "{}\n",
        namespaces_to_sdl(&schema.namespaces, level)
    ));

    result
}
