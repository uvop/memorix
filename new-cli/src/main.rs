mod export_schemas;
mod flat_schema;
mod imports;
mod parser;
mod parser_tools;

use std::error::Error;
use std::fmt;
use std::fs::File;
use std::io::{self, Read, Write};
use std::path::PathBuf;

use nom::error::convert_error;
use nom::Finish;
use parser::{schema_to_sdl, Schema};
use parser_tools::FromSdl;

pub struct PrettyError<T>(pub T);

impl<T> From<T> for PrettyError<T> {
    fn from(v: T) -> Self {
        Self(v)
    }
}

impl<T: fmt::Display> fmt::Debug for PrettyError<T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Display::fmt(&self.0, f)
    }
}

impl From<Box<dyn Error>> for PrettyError<String> {
    fn from(error: Box<dyn Error>) -> Self {
        PrettyError(error.to_string())
    }
}

impl<T: fmt::Debug + fmt::Display> Error for PrettyError<T> {}

impl<T: fmt::Display> fmt::Display for PrettyError<T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.0)
    }
}

pub trait FileSystem {
    fn read_to_string(&self, path: &str) -> io::Result<String>;
    fn write_string(&self, path: &str, content: &str) -> io::Result<()>;
    fn resolve(
        &self,
        relative_file: &str,
        target_file: &str,
    ) -> Result<String, Box<dyn std::error::Error>>;
}

pub struct RealFileSystem;

impl FileSystem for RealFileSystem {
    fn read_to_string(&self, path: &str) -> io::Result<String> {
        let mut file = File::open(path)?;
        let mut contents = String::new();
        file.read_to_string(&mut contents)?;
        Ok(contents)
    }
    fn write_string(&self, path: &str, content: &str) -> io::Result<()> {
        let mut file = std::fs::OpenOptions::new()
            .write(true)
            .create(true)
            .truncate(true)
            .open(path)?;
        file.write_all(content.as_bytes())?;
        Ok(())
    }
    fn resolve(
        &self,
        relative_file: &str,
        target_file: &str,
    ) -> Result<String, Box<dyn std::error::Error>> {
        let path = PathBuf::from(relative_file);
        let new_path = path
            .parent()
            .ok_or("Couldn't get parent of schema".to_string())?
            .join(target_file);
        let abs = new_path.canonicalize()?;
        Ok(abs
            .to_str()
            .ok_or("Couldn't convert path to string")?
            .to_string())
    }
}

pub fn format<F: FileSystem>(
    fs: &F,
    path: &str,
) -> Result<(), Box<dyn std::error::Error + Sync + Send>> {
    let input = fs.read_to_string(path)?;
    let input = input.as_str();
    let (_, schema) = Schema::from_sdl(input).finish().map_err(|e| {
        let stack = format!("parser feedback:\n{}", convert_error(input, e));
        stack
    })?;
    let formatted_input = schema_to_sdl(&schema);
    fs.write_string(path, &formatted_input)?;
    Ok(())
}

pub fn codegen<F: FileSystem>(fs: &F, path: &str) -> Result<String, PrettyError<String>> {
    let parsed_schema =
        crate::imports::ImportedSchema::new(fs, path).map_err(|x| format!("{}", x))?;
    let export_schemas = crate::export_schemas::ExportSchema::new_vec(parsed_schema);
    let flat_export_schemas = export_schemas
        .into_iter()
        .map(crate::flat_schema::FlatExportSchema::new)
        .collect::<Vec<_>>();
    Ok(serde_json::to_string_pretty(&flat_export_schemas).unwrap())
}

enum Command {
    Format,
    Codegen,
}

fn main() -> Result<(), Box<dyn std::error::Error + Sync + Send>> {
    let args: Vec<String> = std::env::args().collect();
    if args.len() != 3 {
        return Err(format!("Usage: [codegen/format] <path_to_sdl_file>").into());
    }
    let command = match args[1].as_str() {
        "format" => Ok(Command::Format),
        "codegen" => Ok(Command::Codegen),
        x => Err(format!("Unknown command \"{}\"", x)),
    }?;

    let file_path = &args[2];
    let path = PathBuf::from(file_path)
        .canonicalize()
        .map_err(|_| "Couldn't get abs path for schema".to_string())?;
    let abs_file_path = path
        .to_str()
        .ok_or("Couldn't get abs path for schema".to_string())?;

    let fs = RealFileSystem {};

    match command {
        Command::Format => format(&fs, abs_file_path)?,
        Command::Codegen => {
            let code = codegen(&fs, abs_file_path)?;
            println!("{}", code);
        }
    }

    Ok(())
}

#[cfg(test)]
mod tests {
    use std::{cell::RefCell, collections::HashMap, io};

    use crate::{codegen, format, FileSystem};

    pub struct MockFileSystem {
        pub files: RefCell<HashMap<String, String>>,
    }

    impl FileSystem for MockFileSystem {
        fn read_to_string(&self, path: &str) -> io::Result<String> {
            self.files
                .borrow()
                .get(path)
                .cloned()
                .ok_or_else(|| io::Error::new(io::ErrorKind::NotFound, "File not found"))
        }

        fn write_string(&self, path: &str, content: &str) -> io::Result<()> {
            self.files
                .borrow_mut()
                .insert(path.to_string(), content.to_string());
            Ok(())
        }

        fn resolve(
            &self,
            _: &str,
            target_file: &str,
        ) -> Result<String, Box<dyn std::error::Error>> {
            Ok(target_file.to_string())
        }
    }

    #[test]
    fn test_parse_example_sdl() -> Result<(), Box<dyn std::error::Error + Sync + Send>> {
        let mock_fs = MockFileSystem {
            files: RefCell::new(HashMap::from([
                (
                    "another-schema.memorix".to_string(),
                    r#"
Config {
  export: {
    engine: Redis( env(REDIS_URL) )
    files: [
      {
        language: typescript
        path: "memorix.generated.ts"
      }
      {
        language: python
        path: "memorix_generated.py"
      }
      {
        language: rust
        path: "memorix_generated.rs"
      }
    ]
  }
}
Type { abc: u32 }
Cache {
  number_of_messages: {
    payload: u32
    public: [get]
  }
  messages: {
    key: string
    payload: string
  }
}
"#
                    .to_string(),
                ),
                (
                    "schema.memorix".to_string(),
                    r#"
Config {
  import: [
    "another-schema.memorix"
  ]
  export: {
    engine: Redis( env(REDIS_URL) )
    files: [
      {
        language: typescript
        path: "memorix.generated.ts"
      }
      {
        language: python
        path: "memorix_generated.py"
      }
      {
        language: rust
        path: "memorix_generated.rs"
      }
    ]
  }
}

Namespace MessageService {
    Enum {
      Operation {
        START
        STOP
      }
    }
    Cache {
        message: {
            key: string
            payload: {
            id: string
            sender_id: UserId
            recipient_id: UserId
            content: string
            timestamp: u64
            }
            public: [get]
        }
    }

    PubSub {
        new_message: {
            key: UserId
            payload: {
            message_id: string
            recipient_id: UserId
            }
            public: [subscribe]
        }
    }

    Task {
        message_processing_tasks: {
            payload: {
                message_id: string
                processing_type: string
                priority: u32
            }
            queue_type: "Fifo"
        }
    }
}

NamespaceDefaults {
  cache_ttl: "3600"
}

Cache {
  user_profile: {
    key: u32
    payload: {
      id: u32
      name: string
      email: string
    }
    public: [get]
  }
  user_session: {
    key: string
    payload: {
      user_id: u32
      session_token: string
      expiry: u64
    }
    ttl: env(USER_SESSION_TTL)
  }
}
  
Namespace UserService {
  NamespaceDefaults {
    cache_ttl: "7200"
    task_queue_type: env(USER_SERVICE_TASK_QUEUE_TYPE)
  }

  Type {
    UserId: u64
  }

  Cache {
    user_profile: {
      key: UserId
      payload: {
        id: UserId
        name: string
        email: string
      }
      public: [get]
    }
    user_session: {
      key: string
      payload: {
        user_id: UserId
        session_token: string
        expiry: u64
      }
      ttl: env(USER_SESSION_TTL)
    }
  }

  PubSub {
    user_activity: {
      key: UserId
      payload: {
        user_id: UserId
        action: string
        timestamp: u64
      }
      public: [subscribe]
    }
  }

  Task {
    user_registration_tasks: {
      payload: {
        user_id: UserId
        email: string
        registration_date: u64
      }
      public: [enqueue get_len]
    }
  }

}

"#
                    .to_string(),
                ),
            ])),
        };
        insta::assert_snapshot!(codegen(&mock_fs, "schema.memorix")?);

        Ok(())
    }

    #[test]
    fn test_fail_nicely() -> Result<(), Box<dyn std::error::Error>> {
        let mock_fs = MockFileSystem {
            files: RefCell::new(HashMap::from([(
                "schema.memorix".to_string(),
                r#"
Config {
  export: {
    engine: Redis(env(REDIS_URL))
    files: [      {
        language: typescript
        path: "memorix.generated.ts"
      }]
  }
}

Type {
  a: [u64 u32]
}

"#
                .to_string(),
            )])),
        };

        let result = codegen(&mock_fs, "schema.memorix");
        let err = result.unwrap_err();
        insta::assert_snapshot!(err.to_string());

        Ok(())
    }
    #[test]
    fn test_format() -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        let mock_fs = MockFileSystem {
            files: RefCell::new(HashMap::from([(
                "schema.memorix".to_string(),
                r#"
    Config {
      export: {
        engine: Redis(env(REDIS_URL))
        files: [      {
            language: typescript
            path: "memorix.generated.ts"
          }]
      }
    }
    Namespace Science {
      Cache {
        how_many_atoms: {
          payload: u32
          key: u64
        }
      }
      PubSub {
        how_many_atoms: {
          payload: u32
        }
      }
    }
    Type {
      a: [u64      ]
    }
    Namespace Rocket {
      Cache {
        launched: {
          payload: boolean
        }
      }
      Enum {
        Operation {
          START
          STOP
        }
        Color {
          RED
          GREEN
          BLUE
        }
      }
      PubSub {
        launched: {
          payload: boolean
        }
      }
    }

    "#
                .to_string(),
            )])),
        };
        format(&mock_fs, "schema.memorix")?;
        let formatted_schema = mock_fs.read_to_string("schema.memorix")?;
        insta::assert_snapshot!(formatted_schema);

        Ok(())
    }
}
