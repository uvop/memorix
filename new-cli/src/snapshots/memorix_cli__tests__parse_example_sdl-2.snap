---
source: src/main.rs
assertion_line: 426
expression: python_code
---
python
